### ISSUE TITLE
[competitor-port] Reinforcement Learning for Trading: FinRL / Stable Baselines3 patterns - Gym Env, PPO/SAC/TD3, Ensemble, Rolling Training

### ISSUE THREAD

**Issue by KCNyu** [KCNyu/clawock#1168]
## 竞品做法
**FinRL** — `finrl/agents/stablebaselines3/models.py` + `finrl/meta/env_stock_trading/env_stocktrading.py`（源码：https://github.com/AI4Finance-Foundation/FinRL/tree/master/finrl）

#### 1. DRLAgent 统一训练/推理接口
- **get_model()**：支持 A2C、DDPG、TD3、SAC、PPO，自动配置超参、action noise
- **train_model()**：统一训练入口，支持 Tensorboard 回调、自定义 callback
- **DRL_prediction()**：统一推理接口，支持确定性/随机策略、资产记忆/动作记忆记录
- **DRL_prediction_load_from_file()**：模型加载与推理一体化

#### 2. DRLEnsembleAgent 多模型集成
- 支持 A2C/PPO/DDPG/SAC/TD3 5 大算法并行训练
- 基于验证集 Sharpe Ratio 动态选择最优模型
- 滚动窗口重训练：`rebalance_window` + `validation_window`
- 动态 turbulence threshold：基于历史波动率分位数自适应调整

#### 3. StockTradingEnv 标准 Gym 环境
- **状态空间**：现金 + 持仓市值 + 技术指标 + 风险指标
- **动作空间**：`Box(-1, 1, shape=(n_assets,))` 连续动作
- **交易逻辑**：
  - 买入：`available_amount = cash // (price * (1 + buy_cost_pct))`
  - 卖出：`sell_num_shares = min(abs(action), holding_shares)`
  - 成本模型：买入/卖出分别扣除 `buy_cost_pct` / `sell_cost_pct`
  - Turbulence 风控：超过阈值强制清仓、禁止买入
- **Reward 设计**：`reward = (total_asset - prev_total_asset) * reward_scaling`
- **终止条件**：遍历完所有交易日
- **输出**：资产净值曲线、每日收益、Sharpe、动作记录、持仓记录

#### 4. 滚动训练/验证/实盘流水线
- **train** → **validation** (Sharpe 选模) → **retrain** (全量数据) → **trade**
- 滚动窗口：`rebalance_window` (训练周期) + `validation_window` (验证周期)
- Ensemble：每个窗口选 Sharpe 最高的模型实盘，状态传递

**Stable Baselines3** — `stable_baselines3/`（源码：https://github.com/DLR-RM/stable-baselines3/tree/master/stable_baselines3）

### 核心算法
| 算法 | 类型 | 适用场景 |
|------|------|----------|
| PPO | On-policy | 连续/离散动作，稳定性强 |
| SAC | Off-policy | 连续动作，最大熵探索 |
| TD3 | Off-policy | 连续动作，Twin Delayed 降低过估计 |
| DDPG | Off-policy | 连续动作，Deterministic Policy Gradient |
| A2C | On-policy | 同步 Actor-Critic，训练快 |

### 核心抽象
- **BaseAlgorithm**：`learn()` / `predict()` / `save()` / `load()`
- **BaseCallback**：`_on_step()` / `_on_rollout_end()` / `_on_training_end()`
- **VecEnv**：`DummyVecEnv` / `SubprocVecEnv` 并行环境
- **ReplayBuffer** / **RolloutBuffer**：经验回放 / 轨迹缓存
- **Policy**：`MlpPolicy` / `CnnPolicy` / `MultiInputPolicy`

**Gymnasium** — 标准 RL 环境接口（源码：https://github.com/Farama-Foundation/Gymnasium）

### 环境接口标准
- `reset(seed=None, options=None)` → `(obs, info)`
- `step(action)` → `(obs, reward, terminated, truncated, info)`
- `observation_space` / `action_space`：`spaces.Box` / `spaces.Discrete` / `spaces.Dict`
- `render()` / `close()`

## 我们现在怎么做的
`src/clawock/portfolio/risk.py` — 仅静态风险度量，**无 RL 智能体、无交易环境、无策略学习**
`src/clawock/portfolio/guardrail.py` — 硬编码规则减仓，**无自适应策略、无探索/利用权衡**
`src/clawock/evaluation/add_alpha_walkforward.py` — 固定因子回测，**无策略学习、无动态仓位优化**
`src/clawock/portfolio/construction.py` — 计划中的组合优化，**无 RL 端到端训练**
`src/clawock/evaluation/add_alpha_walkforward.py:497-512` — 固定仓位回测，**无动态仓位、无交易成本感知策略**

## 差在哪
**机制缺失**：
1. **无 RL 交易环境** — 无标准 Gym 环境、无状态/动作空间定义、无奖励函数设计
2. **无 RL 算法集成** — 无 PPO/SAC/TD3/DDPG/A2C 集成、无超参数管理、无并行环境
3. **无统一训练/推理管线** — 无 `DRLAgent` 风格的统一接口、无模型保存/加载/版本管理
4. **无 Ensemble / 模型选择** — 无验证集指标选模、无滚动重训练、无多模型融合
5. **无风控感知 RL** — 无 turbulence threshold、无成本感知奖励、无强制清仓机制
6. **无回测/实盘一体化** — 训练/验证/实盘流水线割裂、无状态传递、无模型热加载

## 改法
新建/改动文件（≤600 LOC，纯离线核心逻辑，无新增外部服务、无新增 cron）：

1. **新建 `src/clawock/rl/env.py`**（~150 LOC）
   - `TradingEnv` 继承 `gym.Env`：标准股票交易环境
   - 状态：`[cash, holdings_1...n, price_1...n, tech_indicators..., turbulence]`
   - 动作：`Box(-1, 1, (n_assets,))` 连续仓位目标
   - 奖励：`Δtotal_asset * reward_scaling - cost_penalty - risk_penalty`
   - 成本模型：`buy_cost_pct` / `sell_cost_pct` / `slippage_bps`
   - Turbulence 风控：`turbulence_threshold` 超过 → 强制平仓 + 禁买
   - `reset()` / `step()` / `render()` / `save_asset_memory()` / `save_action_memory()`

2. **新建 `src/clawock/rl/agents.py`**（~180 LOC）
   - `DRLAgent` 类：统一 `get_model()` / `train_model()` / `predict()` / `load_model()`
   - 支持算法：`PPO` / `SAC` / `TD3` / `DDPG` / `A2C`（可选依赖 `stable-baselines3`）
   - 超参数注册表：`MODEL_KWARGS`、`NOISE_CONFIG`
   - `TensorboardCallback`：奖励/损失/动作分布记录
   - `DRLAgent.get_model(name, env, **kwargs)` → SB3 模型实例
   - `DRLAgent.train_model(model, total_timesteps, callbacks)` 
   - `DRLAgent.predict(model, env, deterministic=True)` → `(account_history, action_history)`

3. **新建 `src/clawock/rl/ensemble.py`**（~120 LOC）
   - `RLEnsembleAgent`：多算法并行 + 验证集选模
   - `RollingTrainer`：滚动窗口 `rebalance_window` + `validation_window`
   - 动态模型选择：验证集 Sharpe 最高者入选
   - `TurbulenceAdapter`：历史波动率分位数自适应阈值
   - 状态传递：`last_state` 在重训练/实盘间传递

4. **新建 `src/clawock/rl/reward.py`**（~80 LOC）
   - `RewardFunction` Protocol：`compute(state, action, next_state, info) -> float`
   - `SharpeReward`：`(Δasset/asset) / σ(Δasset) * √252`
   - `SortinoReward`：下行风险调整
   - `RiskAdjustedReward`：`return - λ * max_drawdown - μ * turnover`
   - `CostAwareReward`：扣除佣金/滑点/市场冲击

4. **新建 `src/clawock/rl/callbacks.py`**（~60 LOC）
   - `EvalCallback`：定期验证集评估、最优模型保存
   - `EarlyStoppingCallback`：验证指标停滞提前停止
   - `CheckpointCallback`：定期保存模型快照
   - `TurbulenceCallback`：实时监控 turbulence、触发风控

5. **新建 `config/rl_config.json`**（~40 LOC）
```json
{
  "default_algorithm": "ppo",
  "algorithms": {
    "ppo": {"policy": "MlpPolicy", "learning_rate": 3e-4, "n_steps": 2048, "batch_size": 64, "n_epochs": 10, "gamma": 0.99, "gae_lambda": 0.95, "clip_range": 0.2},
    "sac": {"policy": "MlpPolicy", "learning_rate": 3e-4, "buffer_size": 1000000, "batch_size": 256, "tau": 0.005, "gamma": 0.99, "train_freq": 1, "gradient_steps": 1},
    "td3": {"policy": "MlpPolicy", "learning_rate": 1e-3, "buffer_size": 1000000, "batch_size": 256, "tau": 0.005, "gamma": 0.99, "policy_delay": 2},
    "ddpg": {"policy": "MlpPolicy", "learning_rate": 1e-3, "buffer_size": 1000000, "batch_size": 256, "tau": 0.005, "gamma": 0.99},
    "a2c": {"policy": "MlpPolicy", "learning_rate": 7e-4, "n_steps": 5, "gamma": 0.99, "gae_lambda": 1.0}
  },
  "env": {
    "reward_scaling": 1e-4,
    "buy_cost_pct": 0.0005,
    "sell_cost_pct": 0.0005,
    "slippage_bps": 5,
    "turbulence_threshold": "auto",
    "turbulence_lookback": 63,
    "turbulence_quantile": 0.90,
    "max_position_per_asset": 0.3,
    "max_leverage": 1.0
  },
  "training": {
    "total_timesteps": 100000,
    "rebalance_window": 63,
    "validation_window": 21,
    "ensemble_algorithms": ["ppo", "sac", "td3"],
    "selection_metric": "sharpe",
    "tensorboard_log": true
  }
}
```

6. **改 `src/clawock/evaluation/add_alpha_walkforward.py`**（~60 LOC 增量）
   - 新增 `--rl-mode` / `--rl-algorithm` 参数
   - 评估管线可选接入 `RLEnsembleAgent`：训练/验证/实盘一体化
   - 输出 RL 资产曲线、动作分布、Sharpe 对比到 run card

**规模量级**：~630 LOC 新增/改动，纯 Python，依赖 `gymnasium` + 可选 `stable-baselines3`（`pyproject.toml` `extras_require`）。

## 验收
今天会失败、改完会通过的 pytest 断言（贴在测试文件 `tests/test_rl_trading.py`）：

```python
def test_trading_env_reset_step():
    from clawock.rl.env import TradingEnv
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    dates = pd.date_range("2023-01-01", periods=100, freq="B")
    n_assets = 3
    df = pd.DataFrame({
        "date": np.repeat(dates, n_assets),
        "tic": np.tile([f"STK{i}" for i in range(n_assets)], 100),
        "close": np.random.lognormal(4.6, 0.02, 300),
        "volume": np.random.lognormal(13, 0.5, 300),
        "macd": np.random.randn(300),
        "rsi": np.random.uniform(20, 80, 300),
    })
    env = TradingEnv(
        df=df,
        stock_dim=n_assets,
        hmax=100,
        initial_amount=100000,
        buy_cost_pct=[0.001]*n_assets,
        sell_cost_pct=[0.001]*n_assets,
        reward_scaling=1e-4,
        state_space=3*n_assets + 2,
        action_space=n_assets,
        tech_indicator_list=["macd", "rsi"],
    )
    obs, info = env.reset()
    assert obs.shape == (3*n_assets + 2,)
    action = np.array([0.1, -0.1, 0.05])
    next_obs, reward, terminated, truncated, info = env.step(action)
    assert next_obs.shape == obs.shape
    assert isinstance(reward, float)
    assert "total_asset" in info

def test_trading_env_turbulence_control():
    from clawock.rl.env import TradingEnv
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    dates = pd.date_range("2023-01-01", periods=20, freq="B")
    df = pd.DataFrame({
        "date": np.repeat(dates, 2),
        "tic": np.tile(["A", "B"], 20),
        "close": np.random.uniform(10, 100, 40),
        "volume": np.random.lognormal(13, 0.5, 40),
        "turbulence": np.repeat(np.linspace(0, 2, 20), 2),
    })
    env = TradingEnv(
        df=df, stock_dim=2, hmax=100, initial_amount=100000,
        buy_cost_pct=[0.001, 0.001], sell_cost_pct=[0.001, 0.001],
        reward_scaling=1e-4, state_space=8, action_space=2,
        tech_indicator_list=[], turbulence_threshold=1.0,
    )
    env.reset()
    # 强制 turbulence 超过阈值
    env.turbulence = 1.5
    action = np.array([1.0, 1.0])  # 试图买入
    _, _, _, _, info = env.step(action)
    # turbulence 过高应禁止买入
    assert info.get("buy_blocked", False) is True

def test_drl_agent_get_model():
    from clawock.rl.agents import DRLAgent
    import gymnasium as gym
    env = gym.make("CartPole-v1")
    agent = DRLAgent(env)
    model = agent.get_model("ppo", policy="MlpPolicy")
    assert hasattr(model, "learn")
    assert hasattr(model, "predict")

def test_drl_agent_train_predict():
    from clawock.rl.agents import DRLAgent
    import gymnasium as gym
    env = gym.make("CartPole-v1")
    agent = DRLAgent(env)
    model = agent.get_model("ppo", model_kwargs={"n_steps": 64, "batch_size": 32})
    model = agent.train_model(model, tb_log_name="test", total_timesteps=500)
    account, actions = agent.predict(model, env, deterministic=True)
    assert isinstance(account, list)
    assert isinstance(actions, list)

def test_ensemble_agent_model_selection():
    from clawock.rl.ensemble import RLEnsembleAgent
    import numpy as np
    # 模拟验证集 Sharpe
    sharpe_dict = {"ppo": 1.5, "sac": 1.2, "td3": 1.8}
    best = RLEnsembleAgent.select_best_model(sharpe_dict)
    assert best == "td3"

def test_rolling_trainer_window():
    from clawock.rl.ensemble import RollingTrainer
    import pandas as pd
    dates = pd.date_range("2023-01-01", periods=200, freq="B")
    df = pd.DataFrame({"date": dates, "close": np.random.lognormal(4.6, 0.01, 200)})
    trainer = RollingTrainer(rebalance_window=63, validation_window=21)
    windows = list(trainer.generate_windows(df))
    assert len(windows) > 0
    for train_df, val_df, test_df in windows:
        assert len(train_df) >= 63
        assert len(val_df) >= 21

def test_reward_function_sharpe():
    from clawock.rl.reward import SharpeReward
    import numpy as np
    reward_fn = SharpeReward(reward_scaling=1e-4)
    state = {"total_asset": 100000}
    next_state = {"total_asset": 100500}
    reward = reward_fn.compute(state, action=np.array([0.1]), next_state=next_state, info={})
    assert reward > 0

def test_turbulence_adapter():
    from clawock.rl.ensemble import TurbulenceAdapter
    import numpy as np
    np.random.seed(42)
    turbulence = np.random.randn(100).cumsum() + 100
    adapter = TurbulenceAdapter(lookback=20, quantile=0.90)
    threshold = adapter.compute_threshold(turbulence[-30:])
    assert threshold > 0
    assert adapter.is_turbulent(turbulence[-1], threshold) in (True, False)
```

## ROI 与风险
- **影响面**：为组合管理提供**端到端 RL 解决方案**；**某个已发布的数字会更准**（自适应仓位、成本/风险感知奖励、集成模型鲁棒性替代单一因子模型）；接入 #1155 组合构建、#1160 机制检测、#1167 因果推断
- **高压线检查**：
  - live 每日流水线模型输出契约：**不碰** — 仅离线训练库，`add_alpha_walkforward` 可选接入
  - 预注册研究契约 (`peer_residuals.load_rule_config`)：**不碰** — 现有因子/信号保持不变，可作为 RL 状态特征输入
  - dashboard 200KB payload 闸：**不碰** — RL 资产曲线仅在评估/研究内部
- **风险**：RL 训练不稳定（需超参敏感性分析、多种子平均）；奖励函数设计极难（需领域知识）；样本效率低（需离线 RL / 模型-based RL 后续）；过拟合风险（需 Walk-forward 验证、正则化）

**Comment by KCNyu** [comment]
## 关闭理由:产品边界之外(不是"以后再说")

本条由 `competitor-port` 自治循环生成,格式合规、机制描述属实,但它要移植的能力**落在 clawock 的产品定义之外**。

clawock 的终态是 **agent-native 决策工作流插件包 + 可验证 harness**:被外部 agent runtime
(OpenClaw / Claude Code / Codex …) 调用,负责「把证据和对立观点逼进一次决策 / 确定性对账 /
追踪 thesis-decision-execution-outcome / 把已量到的结果变成有界、可评审、可回滚的工作流改进」。
它明确**不做**模型客户端、不做 agent runtime、不做交易执行/做市/清算/合规申报基础设施。

本条要求的能力属于以下之一,均非上述边界内:

- **执行/微结构/做市/SOR/TCA/结算成本** — clawock 不下单、没有 orderbook、没有 venue 连接,
  移植过来的模型没有任何真实输入可喂,只能拿合成数据自证。
- **加密/DeFi/期权/期货/FX 等新资产类** — 需要数据适配器 + 每类的结算守恒不变量 + 重新预注册
  研究契约,是产品边界决策而不是代码任务(同 #1114 留开的那半)。
- **合规申报 / RegTech / KYC-AML** — 面向受监管机构的申报义务,clawock 没有这个身份。
- **重型 ML 基础设施(基础模型/GNN/RL/MLOps/特征商店/流处理/分布式回测)** — 引入的依赖体量与
  运行形态和「装进任何 agent 的插件包」直接冲突;本仓库的研究闸一贯是**样本量与预注册**,
  不是模型容量,新增一个可训练模型只会多一个过拟合面。

关掉不等于否认机制价值。同族里**在边界内**的那几条已单独留着并在实现:回测严谨性
(purged CSCV + PBO,`evaluation/cscv.py` 已上线)、可复现性指纹、因子/信号截面评估
(`clawock signal-panel`)。如果将来产品边界真的扩到某个资产类或执行层,重开一条**带当时的数据
适配器与守恒不变量方案**的 issue,比复活这条更省事。

参考:`clawock-product-goal`(2026-08-08 与 kcn 核对的终态定义)。
